proof/invariant-abstract/DetSchedSchedule_AI.thy

```diff
@@ -1502,30 +1504,22 @@ lemma valid_release_q_no_sc_not_in_release_q:
   "\<lbrakk>valid_release_q s; pred_map_eq None (tcb_scps_of s) ref\<rbrakk> \<Longrightarrow> not_in_release_q ref s"
   by (auto simp: valid_release_q_def in_release_q_def vs_all_heap_simps)
 
-lemma set_tcb_sched_context_valid_release_q_no_sc:
-  "\<lbrace>\<lambda>s. valid_release_q s \<and> pred_map_eq None (tcb_scps_of s) ref\<rbrace>
-   set_tcb_obj_ref tcb_sched_context_update ref sc
-   \<lbrace>\<lambda>_. valid_release_q\<rbrace>"
-  by (wpsimp wp: set_tcb_sched_context_valid_release_q_not_queued
-           simp: valid_release_q_no_sc_not_in_release_q)
-
-lemma set_tcb_sched_context_valid_ready_qs_Some: (* when ref is in ready_queeus *)
-  "\<lbrace>\<lambda>s. valid_ready_qs s
-        \<and> (in_ready_q ref s \<longrightarrow> released_sc_at sp s)\<rbrace>
-   set_tcb_obj_ref tcb_sched_context_update ref (Some sp)
-   \<lbrace>\<lambda>_. valid_ready_qs\<rbrace>"
-  by (wpsimp wp: valid_sched_wp) (auto simp: valid_ready_qs_def in_ready_q_def vs_all_heap_simps)
-
-lemma set_tcb_sched_context_valid_release_q_Some: (* when ref is in release_queue *)
+lemma set_tcb_sched_context_valid_release_q:
   "\<lbrace>\<lambda>s. valid_release_q s
-        \<and> (in_release_q ref s \<longrightarrow> is_active_sc sp s
-                                  \<and> sc_ready_times_of s sp = tcb_ready_times_of s ref) \<rbrace>
-   set_tcb_obj_ref tcb_sched_context_update ref (Some sp)
+        \<and> (in_release_q ref s \<longrightarrow>(\<exists>sp. scopt = Some sp \<and> is_active_sc sp s
+                                       \<and> sc_ready_times_of s sp = tcb_ready_times_of s ref)) \<rbrace>
+   set_tcb_obj_ref tcb_sched_context_update ref scopt
    \<lbrace>\<lambda>_. valid_release_q\<rbrace>"
-  apply (wpsimp wp: valid_sched_wp)
-  apply (clarsimp simp: valid_release_q_def vs_all_heap_simps in_release_q_def sc_ready_time_eq_iff
-                 elim!: sorted_release_qE)
-  by (auto simp: tcb_ready_times_defs map_project_simps opt_map_simps map_join_simps)
+  unfolding set_tcb_obj_ref_def
+  apply simp
+  apply (rule hoare_seq_ext[OF _ gets_the_sp])
+  apply (wpsimp simp: vs_all_heap_simps obj_at_def set_object_def
+                  wp: get_object_wp split: option.splits)
+  by (cases scopt;
```

Comment:
These big blocks always scare me. When they break you don't immediately know where it broke. Is it possble to replace the bit from `rule conjI` with a `fastforce` of some kind?

Reply:
This is not too bad if you know what `valid_release_q` consists of. But it is a bit fiddly to leave out the `sorted_release_q` part at the end. I can certainly pull it apart by repeating things...

Reply:
well, it wasn't actually too bad to make it a bit simpler, although I do think we need two steps, not just one fastforce.